prost-build/src/context.rs

```diff
@@ -0,0 +1,285 @@
+        {
+            return true;
+        }
+        let config_path = match oneof {
+            None => Cow::Borrowed(fq_message_name),
+            Some(oneof_name) => Cow::Owned(format!("{fq_message_name}.{oneof_name}")),
+        };
+        if self
+            .config
+            .boxed
+            .get_first_field(&config_path, field.name())
+            .is_some()
+        {
+            if repeated {
+                println!(
+                    "cargo:warning=\
+                    Field X is repeated and manually marked as boxed. \
+                    This is deprecated and support will be removed in a later release"
+                );
+            }
+            return true;
+        }
+        false
+    }
+
+    /// Returns `true` if this message can automatically derive Copy trait.
+    pub fn can_message_derive_copy(&self, fq_message_name: &str) -> bool {
+        assert_eq!(".", &fq_message_name[..1]);
+        self.message_graph
+            .get_message(fq_message_name)
+            .unwrap()
+            .field
+            .iter()
+            .all(|field| self.can_message_field_derive_copy(fq_message_name, field))
+    }
+
+    /// Returns `true` if the type of this message field allows deriving the Copy trait.
+    pub fn can_message_field_derive_copy(
+        &self,
+        fq_message_name: &str,
+        field: &FieldDescriptorProto,
+    ) -> bool {
+        self.can_field_derive_copy_impl(fq_message_name, None, field)
+    }
+
+    /// Returns `true` if the type of this field in a oneof allows deriving the Copy trait.
+    pub fn can_oneof_field_derive_copy(
+        &self,
+        fq_message_name: &str,
+        oneof_name: &str,
+        field: &FieldDescriptorProto,
+    ) -> bool {
+        self.can_field_derive_copy_impl(fq_message_name, Some(oneof_name), field)
+    }
+
+    fn can_field_derive_copy_impl(
+        &self,
+        fq_message_name: &str,
+        oneof: Option<&str>,
+        field: &FieldDescriptorProto,
+    ) -> bool {
+        assert_eq!(".", &fq_message_name[..1]);
+
+        // repeated field cannot derive Copy
+        if field.label() == Label::Repeated {
+            false
+        } else if field.r#type() == Type::Message {
+            // nested and boxed messages cannot derive Copy
+            if self
+                .message_graph
+                .is_nested(field.type_name(), fq_message_name)
+            {
+                return false;
+            }
+            let config_path = match oneof {
+                None => Cow::Borrowed(fq_message_name),
+                Some(oneof_name) => Cow::Owned(format!("{fq_message_name}.{oneof_name}")),
+            };
+            if self
+                .config
+                .boxed
+                .get_first_field(&config_path, field.name())
+                .is_some()
+            {
+                false
+            } else {
+                self.can_message_derive_copy(field.type_name())
+            }
```

Comment:
This logic has changed compared to `MessageGraph::can_field_derive_copy`. I don't expect that in a PR that mentions just to move code.

Reply:
I will add a regression test to prove this implements correct behavior.

Reply:
For comparison, check the logic in [`Context::should_box_oneof_field`](https://github.com/mzabaluev/prost/blob/6bcdc18e013c1cce16cb804c75285e2654ef10b9/prost-build/src/context.rs#L129) which I did not modify from how it works in the baseline.

Reply:
However, this behavior for fields in a oneof is undocumented, and logically, all named fields in a oneof are still part of the enclosing message's field namespace. So the behavior matching for `boxed` field paths in `can_message_derive_copy` may in fact be a more logical one.

Reply:
I agree there could be a problem here and a regression test is useful. However, this PR should not change any behavior. It should only reorganize the code. Please separate the changed behavior into a separate PR so that we can discuss that in isolation.

My primary concern is that this PR should not introduce a breaking change accidentally. Because it is a lot of code changes, it is difficult to review. If additional changes are slipped in, that makes it even more difficult to review.